File: zadavani.py
from flask import Blueprint, request, jsonify, send_file
import sqlite3
import os
from datetime import datetime

zadavani_bp = Blueprint('zadavani', __name__)

# Cesta k databázi
DB_PATH = os.path.join(os.path.dirname(__file__), "vykaz_cinnosti.db")

# Funkce pro připojení k databázi
def get_db():
    try:
        conn = sqlite3.connect(DB_PATH)
        return conn
    except sqlite3.Error as e:
        print(f"Chyba při připojování k databázi: {e}")
        return None

# Endpoint pro zobrazení stránky zadávání
@zadavani_bp.route("/", methods=["GET"])
def zobraz_zadavani_stranku():
    try:
        return send_file(os.path.join(os.path.dirname(__file__), "zadavani.html"))
    except Exception as e:
        print(f"Chyba při načítání stránky zadávání: {e}")
        return "Chyba při načítání stránky zadávání", 500

# Endpoint pro uložení činnosti
@zadavani_bp.route("/", methods=["POST"])
def ulozit_cinnost():
    data = request.get_json()
    uzivatel_id = data.get("uzivatel_id")
    objekt_id = data.get("objekt_id")
    cas = data.get("cas")
    popis = data.get("popis")
    datum = datetime.now().date()

    if not uzivatel_id or not objekt_id or not cas or not popis:
        return jsonify({"error": "Všechny pole jsou povinné"}), 400

    try:
        with get_db() as conn:
            cursor = conn.cursor()
            cursor.execute("""
                INSERT INTO zaznamy (uzivatel_id, objekt_id, cas, popis, datum)
                VALUES (?, ?, ?, ?, ?)
            """, (uzivatel_id, objekt_id, cas, popis, datum))
            conn.commit()
        return jsonify({"status": "Činnost uložena"}), 201
    except sqlite3.Error as e:
        print(f"Chyba při ukládání činnosti: {e}")
        return jsonify({"error": "Chyba při ukládání činnosti"}), 500

# Endpoint pro načtení uživatele a přiřazených objektů, včetně hodin
@zadavani_bp.route("/data/<int:uzivatel_id>", methods=["GET"])
def nacti_data_pro_uzivatele(uzivatel_id):
    try:
        with get_db() as conn:
            cursor = conn.cursor()

            # Načtení informací o uživateli
            cursor.execute("SELECT jmeno FROM uzivatele WHERE id = ?", (uzivatel_id,))
            uzivatel = cursor.fetchone()

            if not uzivatel:
                return jsonify({"error": "Uživatel nenalezen"}), 404

            # Načtení přiřazených objektů a informací o fondech hodin a vyčerpaných hodinách
            cursor.execute("""
                SELECT o.id, o.nazev, o.fond_hodin,
                       IFNULL(SUM(z.cas), 0) AS vycerpane_hodiny
                FROM prirazeni p
                JOIN objekty o ON p.objekt_id = o.id
                LEFT JOIN zaznamy z ON z.objekt_id = o.id AND z.uzivatel_id = p.uzivatel_id
                WHERE p.uzivatel_id = ?
                GROUP BY o.id, o.nazev, o.fond_hodin
            """, (uzivatel_id,))
            objekty = cursor.fetchall()

            return jsonify({
                "uzivatel": {"id": uzivatel_id, "jmeno": uzivatel[0]},
                "objekty": [{"id": o[0], "nazev": o[1], "fond_hodin": o[2], "vycerpane_hodiny": o[3]} for o in objekty]
            })
    except sqlite3.Error as e:
        print(f"Chyba při načítání dat pro uživatele: {e}")
        return jsonify({"error": "Chyba při načítání dat"}), 500

# Endpoint pro poskytování JavaScriptového souboru zadavani.js
@zadavani_bp.route("/zadavani.js", methods=["GET"])
def poskytnout_zadavani_js():
    try:
        return send_file(os.path.join(os.path.dirname(__file__), "zadavani.js"))
    except Exception as e:
        print(f"Chyba při načítání souboru zadavani.js: {e}")
        return "Chyba při načítání souboru zadavani.js", 500

File: zadavani.js
document.addEventListener('DOMContentLoaded', () => {
    console.log("Stránka zadávání byla načtena, volám funkci pro načítání dat...");
    const params = new URLSearchParams(window.location.search);
    const uzivatelId = params.get('uzivatel_id');

    if (!uzivatelId) {
        alert("Chybí ID uživatele v URL!");
        return;
    }

    nactiDataProUzivatele(uzivatelId);
});

function nactiDataProUzivatele(uzivatelId) {
    fetch(`/zadavani/data/${uzivatelId}`)
        .then(response => response.json())
        .then(data => {
            if (data.error) {
                console.error("Chyba při načítání dat:", data.error);
                return;
            }

            // Nastavení jména uživatele
            const uzivatelJmeno = data.uzivatel.jmeno;
            document.getElementById('uzivatelJmeno').textContent = `Uživatel: ${uzivatelJmeno}`;

            // Načtení přiřazených objektů do selectu
            const objektSelect = document.getElementById('objektSelect');
            objektSelect.innerHTML = '<option value="" disabled selected>Vyberte objekt</option>';
            data.objekty.forEach(({ id, nazev }) => {
                const option = document.createElement('option');
                option.value = id;
                option.textContent = nazev;
                objektSelect.appendChild(option);
            });

            // Načtení informací o fondech hodin a zobrazení v tabulce
            const hodinySeznam = document.getElementById('hodinySeznam');
            hodinySeznam.innerHTML = '';
            data.objekty.forEach(({ id, nazev, fond_hodin, vycerpane_hodiny }) => {
                const zbyvaHodin = fond_hodin - vycerpane_hodiny;
                const procentVycerpano = fond_hodin > 0 ? ((vycerpane_hodiny / fond_hodin) * 100).toFixed(2) : 0;

                const listItem = document.createElement('tr');
                listItem.innerHTML = `
                    <td class="p-2 border">${nazev}</td>
                    <td class="p-2 border">${fond_hodin}</td>
                    <td class="p-2 border">${vycerpane_hodiny}</td>
                    <td class="p-2 border">${zbyvaHodin}</td>
                    <td class="p-2 border">${procentVycerpano}%</td>
                `;
                hodinySeznam.appendChild(listItem);
            });
        })
        .catch(error => {
            console.error("Chyba při načítání dat pro uživatele:", error);
        });
}

function ulozitCinnost() {
    console.log("Funkce ulozitCinnost byla zavolána.");

    const params = new URLSearchParams(window.location.search);
    const uzivatelId = params.get('uzivatel_id');

    if (!uzivatelId) {
        alert("Chybí ID uživatele v URL!");
        return;
    }

    const objektId = document.getElementById('objektSelect').value;
    const cas = document.getElementById('cas').value;
    const popis = document.getElementById('popis').value;

    if (!objektId || !cas || !popis) {
        vytvorNotifikaci("Všechny pole jsou povinné!", 'bg-red-500');
        return;
    }

    fetch('/zadavani', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ uzivatel_id: uzivatelId, objekt_id: objektId, cas, popis }),
    })
    .then(response => response.json())
    .then(data => {
        if (data.error) {
            vytvorNotifikaci("Chyba při ukládání činnosti: " + data.error, 'bg-red-500');
        } else {
            vytvorNotifikaci("Činnost úspěšně uložena!", 'bg-green-500');
            document.getElementById('zadavaniForm').reset();
            nactiDataProUzivatele(uzivatelId); // Načtení aktualizovaných dat
        }
    })
    .catch(error => {
        vytvorNotifikaci("Chyba při ukládání činnosti: " + error.message, 'bg-red-500');
    });
}

// Definice funkce pro vytváření notifikací
function vytvorNotifikaci(zprava, barva) {
    const notifikace = document.createElement('div');
    notifikace.className = `p-3 mb-4 text-white rounded ${barva}`;
    notifikace.textContent = zprava;

    const notifikaceKontejner = document.getElementById('notifikaceKontejner');
    notifikaceKontejner.appendChild(notifikace);
    setTimeout(() => {
        notifikace.remove();
    }, 3000);
}
